File: Swift/Sources/BiabKit/Money.swift
import Foundation

/// Currency formatting.
///
/// Two entry points on purpose, because the API is not uniform and guessing
/// wrong is a 100× error in either direction:
///
/// * ``cents(_:currency:)`` — product prices, subscription prices, checkout
///   totals, invoice and quote totals. These are **integer cents**, the
///   platform-wide convention for stored money. It exists so no total picks up
///   a floating-point rounding error on the way to Stripe.
/// * ``amount(_:currency:)`` — cart `unitPrice` / `subtotal`, which arrive
///   already decimal.
///
/// Before rendering a money field this kit doesn't already model, check which
/// shape it is. The model doc comments say so per property.
public enum Money {
    public static func cents(_ value: Int?, currency: String? = "USD") -> String {
        guard let value else { return "" }
        return format(Decimal(value) / 100, currency: currency)
    }

    public static func amount(_ value: Double?, currency: String? = "USD") -> String {
        guard let value else { return "" }
        return format(Decimal(value), currency: currency)
    }

    private static func format(_ value: Decimal, currency: String?) -> String {
        // Decimal, not Double: money formatting should not round-trip through
        // binary floating point.
        value.formatted(.currency(code: currency ?? "USD"))
    }
}

File: Phoenix/lib/bd_starter_web/money.ex
defmodule BdStarterWeb.Money do
  @moduledoc """
  Currency formatting.

  Two functions on purpose, because the API is not uniform and guessing wrong
  is a 100× error in either direction:

    * `cents/2` — product prices, subscription prices, checkout totals. These
      are **integer cents**, the platform-wide convention for stored money. It
      exists so no total picks up a floating-point rounding error on the way
      to Stripe.
    * `amount/2` — cart `unitPrice` / `subtotal`, which arrive already decimal.

  Before rendering a money field this starter doesn't already show, check
  which shape it is.
  """

  # Bodiless heads: defaults can only be declared once when a function has
  # several clauses.
  def cents(value, currency \\ "USD")
  def cents(nil, _currency), do: ""
  def cents(value, currency) when is_number(value), do: format(value / 100, currency)

  def cents(value, currency) when is_binary(value) do
    case Float.parse(value) do
      {n, _} -> format(n / 100, currency)
      :error -> ""
    end
  end

  def cents(_value, _currency), do: ""

  def amount(value, currency \\ "USD")
  def amount(nil, _currency), do: ""
  def amount(value, currency) when is_number(value), do: format(value * 1.0, currency)

  def amount(value, currency) when is_binary(value) do
    case Float.parse(value) do
      {n, _} -> format(n, currency)
      :error -> ""
    end
  end

  def amount(_value, _currency), do: ""

  defp format(value, "USD"), do: "$" <> :erlang.float_to_binary(value * 1.0, decimals: 2)

  defp format(value, currency),
    do: currency <> " " <> :erlang.float_to_binary(value * 1.0, decimals: 2)
end
